tests/sccache_cargo.rs

```diff
@@ -64,74 +112,123 @@ fn test_rust_cargo_cmd(cmd: &str, extra_envs: &[(&str, &std::ffi::OsStr)]) {
             .parse_env("RUST_LOG")
             .try_init(),
     );
-    let cargo = env!("CARGO");
-    debug!("cargo: {}", cargo);
-    let sccache = assert_cmd::cargo::cargo_bin("sccache");
-    debug!("sccache: {:?}", sccache);
-    let crate_dir = Path::new(file!()).parent().unwrap().join("test-crate");
+
+    debug!("cargo: {:?}", CARGO);
+    debug!("sccache: {:?}", SCCACHE_BIN);
     // Ensure there's no existing sccache server running.
-    stop();
-    // Create a temp directory to use for the disk cache.
-    let tempdir = tempfile::Builder::new()
-        .prefix("sccache_test_rust_cargo")
-        .tempdir()
-        .unwrap();
-    let cache_dir = tempdir.path().join("cache");
-    fs::create_dir(&cache_dir).unwrap();
-    let cargo_dir = tempdir.path().join("cargo");
-    fs::create_dir(&cargo_dir).unwrap();
-    // Start a new sccache server.
-    trace!("sccache --start-server");
-    sccache_command()
-        .arg("--start-server")
-        .env("SCCACHE_DIR", &cache_dir)
-        .assert()
-        .success();
+    stop_sccache()?;
+
+    test_rust_cargo_cmd("check", SccacheTest::new(None)?)
+        .context("Sccache failed for `cargo check`")?;
+    test_rust_cargo_cmd("build", SccacheTest::new(None)?)
+        .context("Sccache failed for `cargo build`")?;
+
+    test_rust_cargo_env_dep(SccacheTest::new(None)?)
+        .context("Sccache failed for `cargo build` with changed environment variables.")?;
```

Comment:
Let's split these into three test functions:
* One checking that a base cargo build has cache hits
* One checking that a cargo `-Zprofile` build has cache hits
* One checking that changing a needed env var has a cache miss.

Reply:
I actually tried that out, but there is a complication. Cargo test by default invokes multiple tests in parallel - and this doesn't work well with sccache. We already use different directories, but sccache complains "address already in use" when starting a second sccache server.

So you can only test sequentially. There is an option for cargo test to do that, but by default `cargo test` would result in failing tests. 
So we would have to add a bit of extra complexity - E.g. use a Mutex (but keep lock poisining in mind) or an atomic variable to synchronize access and make sure only one test runs at a time, even if multiple are started in parallel.

Reply:
Thanks for the suggestion - `serial_test` worked just fine.